### PR TITLE
Fill out test plan for auto-default structs

### LINKED ISSUE #60167: Test plan for "auto-default structs"
Spec: dotnet/csharplang#5737

- #60176
- EnC (implicit initializations should occur after sequence point for opening `{` in constructor?)

## Compiler
- [x] add proposal to dotnet/csharplang (dotnet/csharplang#5964)
- [x] add to compiler test plan
- [x] `-langversion`
- [x] field initialization:
  - [x] when all fields are explicitly assigned
  - [x] when no fields are assigned (use `initobj`?)
      - implementation initializes each field individually.
  - [x] when no fields
  - [x] with auto-properties with `get` only; with `set`; with `init`
      - `SemanticErrorTests.CS0843ERR_UnassignedThisAutoProperty`,
      - `FlowDiagnosticTests.AutoPropInitialization4`
      - `StructConstructorTests.FieldInitializers_08`
  - [x] with field-like events
  - [x] when all sub-fields of a field are explicitly assigned
  - [x] when some sub-fields of a field are explicitly assigned
  - [x] when field is conditionally assigned
  - [x] when explicitly calling declared constructor `: this(...)`
  - [x] when explicitly calling default parameterless constructor `: this()`
  - [x] when assigning `this = default;`
  - [x] when assigning `this = default;` conditionally
  - [x] ~~when assigning to backing field of semi-auto property (i.e. using a field initializer)~~ moved to #57012
  - [x] ~~when assigning to property of semi-auto property (i.e. in a constructor)~~ moved to #57012
  - [x] when reading field before assigning
- [x] fields assigned where C#10 produces warnings with `/warn:5`
- [x] `record`
  - [x] with explicit fields
  - [x] with explicit auto-properties
- [x] disabled-by-default diagnostics:
  - [x] when all fields are explicitly assigned
  - [x] when no fields are assigned
  - [x] when diagnostics are elevated to warnings or errors
- [x] `default` assigned:
  - [x] to backing fields of auto-properties
  - [x] for field of type parameter type
  - [x] for field of `struct` type with parameterless constructor
  - [x] for field of empty `struct` type (no assignment?)
- [x] field initialization when parameters have `!!` (which runs first?)
- [x] nullable analysis treats auto-initialized fields as nullable
- [x] `#pragma warning restore` should not enable disabled diagnostics

## Productivity
- [x] Add new error codes to UpgradeProject fixer
  - [ ] upgrade from warnings?
- [x] Debugger: reposition to first sequence point in constructor (init fields?)

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #60167 (test plan)

I'm not sure how to test the LangVersion warnings here, as it requires having a struct with private reference type fields in another assembly, but I did test it manually. [langversion-warning](https://user-images.githubusercontent.com/5833655/163487584-c630a767-e733-4078-ac58-51d9a0bc05be.gif)

I manually verified in the debugger that changing an auto-defaulted field's value, then dragging to the `{` of the constructor, causes default to get written to the field again. [auto-default-debugger](https://user-images.githubusercontent.com/5833655/163485843-6160b0a5-cb09-47c3-a7a0-a2ab221da08e.gif)

### CODE REVIEW

**Comment by RikkiGibson** [comment]
Bleh, the fix for sequence points effectively means I need to change a *lot* of test baselines. Going to draft for now.

**Comment by RikkiGibson** [comment]
@jcouv @cston this is ready for review.

**Comment by cston** [comment]
> I'm not sure how to test the LangVersion warnings here, as it requires having a struct with private reference type fields in another assembly

Is the question how to write an "upgrade project" test with multiple projects? It looks like there are examples such as `UpgradeProjectForSealedToStringInRecords_CS8912`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs [hunk 1]
```diff
@@ -2696,6 +2755,482 @@ .maxstack  2
...
+{
+  // Code size        9 (0x9)
+  .maxstack  1
+  IL_0000:  nop
+  IL_0001:  ldarg.0
+  IL_0002:  initobj    ""S2""
+  IL_0008:  ret
+}
+");
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs:2853]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing `S1 = default;` as well.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs [hunk 2]
```diff
@@ -2696,6 +2755,482 @@ .maxstack  2
...
+}";
+            var comp = CreateCompilation(source, options: TestOptions.DebugDll.WithSpecificDiagnosticOptions(ReportStructInitializationWarnings));
+            comp.VerifyDiagnostics(
+                // (6,12): warning CS9022: Control is returned to caller before field 'S.X' is explicitly assigned, causing a preceding implicit assignment of 'default'.
+                //     public S()
+                Diagnostic(ErrorCode.WRN_UnassignedThisSupportedVersion, "S").WithArguments("S.X").WithLocation(6, 12));
+
+            comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs:3100]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing `pragma warning disable 9022` as well, with and without `WithSpecificDiagnosticOptions(ReportStructInitializationWarnings)`.

**Comment by RikkiGibson** [comment]
> > I'm not sure how to test the LangVersion warnings here, as it requires having a struct with private reference type fields in another assembly
> 
> Is the question how to write an "upgrade project" test with multiple projects? It looks like there are examples such as `UpgradeProjectForSealedToStringInRecords_CS8912`.

I poked at this a bit more but I wasn't able to get the warning wave diagnostic to show up. Not sure why.


```cs
        [Fact]
        public async Task UpgradeProjectForStructAutoDefaultError_4()
        {
            await TestLanguageVersionUpgradedAsync(
@"
<Workspace>
    <Project Language=""C#"" AssemblyName=""Assembly1"" CommonReferences=""true"" LanguageVersion=""10"">
        <Document FilePath=""Derived.cs"">
public struct S1
{
    private object objectField;
}
        </Document>
    </Project>
    <Project Language=""C#"" AssemblyName=""Assembly2"" CommonReferences=""true"" LanguageVersion=""10"">
        <ProjectReference>Assembly1</ProjectReference>
        <Document FilePath=""Base.cs"">
public struct S2
{
    public S1 structField;
    public [|S2|](bool unused) { }
}
        </Document>
    </Project>
</Workspace>
",
                expected: LanguageVersion.Preview,
                new CSharpParseOptions(LanguageVersion.CSharp10),
                // also added as a parameter to 'TestLanguageVersionUpgradedAsync' and threaded through
                new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary, warningLevel: 9999));
```

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 3]
```diff
@@ -1224,14 +1224,13 @@ void addDiagnosticForStructThis(Symbol thisParameter, int thisSlot)
...
                 {
-                    Diagnostics.Add(ErrorCode.WRN_UseDefViolationThisSupportedVersion, node.Location, thisParameter.Name);
+                    Diagnostics.Add(ErrorCode.WRN_UseDefViolationThisSupportedVersion, node.Location);
                 }
                 else
                 {
                     Diagnostics.Add(
                         ErrorCode.ERR_UseDefViolationThisUnsupportedVersion,
                         node.Location,
-                        thisParameter.Name,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1234]: nit: I wonder if our diagnostic infrastructure should/could have caught that, the same way that it verifies that all parts are "printable".

File: src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs [hunk 4]
```diff
@@ -2696,6 +2755,535 @@ .maxstack  2
 ");
         }
 
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(LanguageVersion.Latest)]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs:2760]: nit: probably should be Preview or Next, we don't really use Latest much

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7). Left a couple of comments for nits

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1234]: Perhaps the `DiagnosticInfo` constructor should verify `args.Length - 1` matches the largest index in the message.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1234]: Filed #60862

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-ide for review of IDE-specific changes

File: src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs [hunk 5]
```diff
@@ -1103,5 +1103,44 @@ void M()
...
+        {
+            await TestLanguageVersionUpgradedAsync(@"
+struct Test
+{
+    public int X { get; set; }
+    public [|Test|]() { this.ToString(); }
+}",
+                expected: LanguageVersion.Preview,
+                new CSharpParseOptions(LanguageVersion.CSharp10));
+        }
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs:1144]: 💭 It looks like there are 6 new error cases added, but only 3 tested.

**Review by sharwell — APPROVED** [review]
IDE side looks fine. Would be good to add remaining tests if possible.

File: src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs:1144]: Please see https://github.com/dotnet/roslyn/pull/60688#issuecomment-1104345627

**Pull request merged**